File: DesignPatternsExercises/java/MementoDesignPattern/Solution/HomeTheatre.java
public class HomeTheatre {
	private LedTV ledTV;
	private Speaker speaker;
	private static int stateCount = 0;

	public HomeTheatre(LedTV ledTV, Speaker speaker) {
		this.ledTV = ledTV;
		this.speaker = speaker;
		// __________(1)__________++;
		stateCount++;
	}

	public Memento createMemento() {
		// return __________(2)__________(ledTV, speaker);
		return new Memento(ledTV, speaker);
	}

	public void setMemento(Memento memento) {
		// ledTV = memento.__________(3)__________();
		ledTV = memento.getLedTV();
		// speaker = memento.__________(4)__________();
		speaker = memento.getSpeaker();
	}

	public static int getStateCount() {
		return stateCount;
	}

	@Override
    public String toString() {
		return ledTV + "\n    " + speaker;
    }
}

File: DesignPatternsExercises/ruby/memento_design_pattern/solution/home_theatre.rb
require_relative 'memento'

class HomeTheatre
  @@state_count = 0

  def initialize(led_tv, speaker)
    @led_tv = led_tv
    @speaker = speaker
    # __________(1)__________ += 1
    @@state_count += 1
  end

  # def __________(2)__________
  def self.state_count
    @@state_count
  end

  def memento
    # __________(3)__________(@led_tv, @speaker)
    Memento.new(@led_tv, @speaker)
  end

  def memento=(memento)
    # @led_tv = __________(4)__________
    @led_tv = memento.led_tv
    # @speaker = __________(5)__________
    @speaker = memento.speaker
  end

  def to_string
    "#{@led_tv.to_string}\n    #{@speaker.to_string}"
  end
end
